apps/frontend/src/libs/hooks/use-pagination/use-pagination.ts

```diff
@@ -0,0 +1,75 @@
+import {
+	useCallback,
+	useEffect,
+	useSearchParams,
+	useState,
+} from "~/libs/hooks/hooks.js";
+
+import { DEFAULT_VALUES, FIRST_PAGE } from "./libs/constants/constants.js";
+import { QueryParameter } from "./libs/enums/enums.js";
+import { calculateTotalPages } from "./libs/helpers/helpers.js";
+
+type Parameters = {
+	totalItems: number;
+};
+
+type ReturnType = {
+	onPageChange: (page: number) => void;
+	onPageSizeChange: (pageSize: number) => void;
+	page: number;
+	pageSize: number;
+};
+
+const usePagination = ({ totalItems }: Parameters): ReturnType => {
+	const [searchParameters, setSearchParameters] = useSearchParams();
+
+	// There is no check for page and pageSize to be in available range
+	const pageQueryParameter =
+		Number(searchParameters.get(QueryParameter.page)) || DEFAULT_VALUES.PAGE;
+	const pageSizeQueryParameter =
+		Number(searchParameters.get(QueryParameter.pageSize)) ||
+		DEFAULT_VALUES.PAGE_SIZE;
```

Comment:
Set these properties as initial values in useEffect, and here you will be able to add a check whether the page is in range

Reply:
Not sure, what do you mean (

Reply:
The point is that we should return only valid values from this hook. Here is an example of what I mean:

```typescript
// remove default values and parsing to number here
const pageQueryParameter = searchParameters.get(QueryParameter.page);
const pageSizeQueryParameter = searchParameters.get(QueryParameter.pageSize);

const [page, setPage] = useState(DEFAULT_PAGE);
const [pageSize, setPageSize] = useState(DEFAULT_PAGE_SIZE);

useEffect(() => {
 // parse query params values (turn into numbers)
 // check that values are in range, if not override query params with default values
 // set page and pageSize with valid values

}, [pageQueryParameter, pageSizeQueryParameter])

// return only valid page and pageSize from hook
```

Reply:
I tried to do this and as I was afraid it cause render problem, cause this useEffect changes page/pageSize when queryParams updates, and other useEffect listens for page/pageSize updates and change queryParams. So here is a cyclic problem. But I get the point, and will try to do something

Reply:
I tried to do this way, but there was an issue with select component. Cause when I usePagination hook called, it initializes page/pageSize states with DEFAULT_..., and it not simultaniously execute useEffect. So inside  `TablePagination` component first passed those default values, and useAppForm uses this default values. And after useEffect executes, those values, I passed inside useAppForm are not being updated, and it rendered not correct number. Also cause of useEffect executes not simultaniously, there is some ripple effect in search params (after page refresh, first the values are equals to defaults, and then they change to previous, which were before refresh -- here I expect, that they just stay as was before refresh, and not change for a moment)